File: sdk/mobile/android/src/test/kotlin/com/summit/odps/android/TestFixtures.kt
package com.summit.odps.android

import kotlin.random.Random

class TestNetworkState(var available: Boolean) : NetworkStateProvider {
  override fun isNetworkAvailable(): Boolean = available
}

class TestUploader : TelemetryUploader {
  val batches = mutableListOf<List<TransmissionEnvelope>>()
  var shouldFailOnce: Boolean = false

  override fun upload(batch: List<TransmissionEnvelope>): UploadResult {
    if (shouldFailOnce) {
      shouldFailOnce = false
      return UploadResult(success = false, errorMessage = "simulated failure")
    }
    batches += batch.map { it.copy() }
    return UploadResult(success = true)
  }
}

class SeededLaplaceNoiseGenerator(seed: Int) : NoiseGenerator {
  private val random = Random(seed)
  override fun sample(scale: Double): Double {
    if (scale == 0.0) return 0.0
    val u = random.nextDouble() - 0.5
    val magnitude = kotlin.math.ln(1 - 2 * kotlin.math.abs(u)) * -scale
    return kotlin.math.sign(u) * magnitude
  }
}

File: sdk/mobile/ios/Tests/ODPSTests/TestFixtures.swift
import Foundation
@testable import ODPS

final class TestNetworkState: NetworkStateProvider {
    var available: Bool

    init(available: Bool) {
        self.available = available
    }

    func isNetworkAvailable() -> Bool {
        available
    }
}

final class TestUploader: TelemetryUploader {
    var batches: [[TransmissionEnvelope]] = []
    var shouldFailOnce: Bool = false

    func upload(batch: [TransmissionEnvelope]) -> UploadResult {
        if shouldFailOnce {
            shouldFailOnce = false
            return UploadResult(success: false, errorMessage: "simulated failure")
        }
        batches.append(batch)
        return UploadResult(success: true)
    }
}

func makePolicyPack() -> PolicyPack {
    let eventPolicy = EventPolicy(eventType: "session_start",
                                  directives: [
                                    "sessionId": FieldDirective(action: .allow),
                                    "appVersion": FieldDirective(action: .allow),
                                    "piiEmail": FieldDirective(action: .deny, isSensitive: true)
                                  ])
    let metrics = [
        "view_count": MetricPolicy(name: "view_count", type: .count, epsilon: 1.0, sensitivity: 1.0),
        "duration_sum": MetricPolicy(name: "duration_sum", type: .sum, epsilon: 0.8, sensitivity: 5.0)
    ]
    return PolicyPack(eventPolicies: ["session_start": eventPolicy], metricPolicies: metrics)
}
